google/cloud/spanner/internal/session_pool.cc

```diff
@@ -133,11 +115,103 @@ void SessionPool::ScheduleBackgroundWork(std::chrono::seconds relative_time) {
 }
 
 void SessionPool::DoBackgroundWork() {
+  MaintainPoolSize();
   // TODO(#1171) Implement SessionPool session refresh
-  // TODO(#1172) maintain desired SessionPool size
   ScheduleBackgroundWork(std::chrono::seconds(5));
 }
 
+void SessionPool::MaintainPoolSize() {
+  // Ensure the pool size conforms to what was specified in the
+  // `SessionOptions`, creating or deleting sessions as necessary.
+  std::unique_lock<std::mutex> lk(mu_);
+  if (!create_in_progress_ && total_sessions_ < options_.min_sessions()) {
+    Grow(lk, total_sessions_ - options_.min_sessions(), /*async=*/true);
+  }
+}
+
+/**
+ * Grow the session pool by creating up to `sessions_to_create` sessions and
+ * adding them to the pool.  Note that `lk` may be released and reacquired in
+ * this method.
+ *
+ * TODO(#1271) eliminate the `async` parameter and do all creation
+ * asynchronously. The main obstacle is making existing tests pass.
+ */
+Status SessionPool::Grow(std::unique_lock<std::mutex>& lk,
+                         int sessions_to_create, bool async) {
+  int num_channels = static_cast<int>(channels_.size());
+  int session_limit = options_.max_sessions_per_channel() * num_channels;
+  if (total_sessions_ == session_limit) {
+    // Can't grow the pool since we're already at max size.
+    return Status(StatusCode::kResourceExhausted, "session pool exhausted");
+  }
+
+  create_in_progress_ = true;
+
+  // Compute how many Sessions to create on each Channel, trying to keep the
+  // number of Sessions on each channel equal.
+  //
+  // However, the counts may become unequal over time, and we do not want
+  // to delete sessions just to make the counts equal, so do the best we
+  // can within those constraints.
+  int target_total_sessions =
+      (std::min)(total_sessions_ + sessions_to_create, session_limit);
```

Comment:
fwiw, I don't think the `(std::min)` trick is needed in .cc files, because we know what our includes are and if our build works, then it works. This differes from `.h` files where callers may include us after including the offending Windows include that defines the macro for `min`.